SIMD-0392: Relax post-execution balance check

igor56D opened the pull request:
Implementation of the proposal described in [SIMD-0392](https://github.com/solana-foundation/solana-improvement-documents/blob/main/proposals/0392-rent-increase-adaptations.md). 

#### Summary of Changes

- ~~SIMD-0392 assumes all existing accounts are rent-exempt pre-execution, so the `RentState::RentPaying` state is ignored in several places (most notably in `new_pre_exec` and `validate_fee_payer`). IOW the expectation is that no rent paying accounts exist currently and none can be created, which I believe is already true on mainnet.~~
  - ~~many tests using rent paying accounts and expecting those semantics needed to be modified. A few that specifically tested the processing of rent paying accounts were removed as they should no longer be necessary.~~
  - Moved these changes into a separate [PR](https://github.com/anza-xyz/agave/pull/9438) and rebased this branch onto it.
- Feature gating in `execute_loaded_transaction` to introduce new post-exec balance checking logic. When the feature is activated, any state transitions containing a `RentPaying` rent state (either pre or post exec) will be explicitly disallowed (rather than implicitly, which is the case now).
- The post-exec minimum balance requirement is `min(rent_exempt_minimum(acc), acc.pre_exec_balance)` if the account wasn't allocated or reallocated during execution. Otherwise, the account is subject to the current rent exempt minimum as before (see SIMD for more details). This allows for rent increases without making existing accounts `RentPaying`.
- merged the rent adjusted delegation changes from https://github.com/igor56D/agave/pull/1 into this branch.

#### TODO
- ~~feature gate keypair generation~~

igor56D committed:
initial implementation: SIMD-0392

igor56D commented on runtime/src/bank/fee_distribution.rs:
```diff
@@ -152,28 +152,11 @@ impl Bank {
-            account.lamports(),
-            account.data().len(),
-        );
-        let rent_state_transition_allowed =
```
I don't think this check makes sense with rent-exempt accounts. an existing account is guaranteed to be rent-exempt and depositing additional funds won't change that. correct me if I'm wrong.

igor56D replied:
I added a check to ensure that if the fee receiver account doesn't exist the fee amount must exceed the rent-exempt minimum for the payment to succeed. This should be consistent with upstream behavior.

igor56D commented:
> why did this influence so many tests? it's supposed to be relaxing constraints

I mentioned that in the PR description:

> many tests using rent paying accounts and expecting those semantics needed to be modified. A few that specifically tested the processing of rent paying accounts were removed as they should no longer be necessary.

The most common case was RentPaying accounts, which are no longer allowed on MB and don't make sense in the context of 0392, being used as fee payers or senders. A RentPaying account can have its balance decremented but a rent-exempt account below or at the rent-exempt minimum cannot. For these, I increased balances for the necessary accounts.

A few tests were directly testing state transitions for RentPaying accounts. Those were deleted.

igor56D commented:
> ok but those changes can be done in a preliminary pr

fair enough. All the test changes I mentioned are contained to the last commit on this branch. I'll move that to a separate branch and make a new PR.

igor56D commented:
@t-nelson FYI I created a [separate PR](https://github.com/anza-xyz/agave/pull/9438) for the test changes if you want to take a look.

buffalojoec commented on svm/src/transaction_processor.rs:
```diff
@@ -1049,14 +1059,17 @@ impl<FG: ForkGraph> TransactionBatchProcessor<FG> {
         } = execution_record;
 
-        if status.is_ok()
+        if post_account_state_info.is_ok()
```
Why change this from `status` (`Result<(), TransactionError>`)? It doesn't seem like you need the post-state info after `verify_changes`.

igor56D commented on svm/src/transaction_account_state_info.rs:
```diff
@@ -5,94 +5,181 @@ use {
+                {
+                    // the same account MUST be present and marked writable in both pre and post exec
+                    debug_assert!(pre_exec_state_info.is_some());
+                    let pre_exec_state_info = pre_exec_state_info.as_ref().unwrap();
```
I was on the fence with this. Given that the pre-exec will return `None` instead of panic if borrow fails, it doesn't make sense to then panic on that `None` here instead of where the issue originated. I'll change.

igor56D commented on svm/src/transaction_account_state_info.rs:
```diff
@@ -5,94 +5,181 @@ use {
+                    // otherwise, pre-exec balance can also be used as the minimum balance
+                    let mut min_balance = rent.minimum_balance(account.data().len());
+                    if pre_exec_state_info.rent_state != RentState::Uninitialized
+                        && account.data().len() <= pre_exec_state_info.data_size
```
Yes, good catch! I wrote this before that change was made to the SIMD and forgot to update.

buffalojoec reviewed (commented):
Looks like my previous comments were resolved, thanks @igor56D!

I guess the main piece left is ironing out exactly what role `RentPaying` plays here and how to manage that rent state.

I'm also wondering if maybe we should bolster this change with some integration tests - maybe ones that run some transactions and modify lamports, state, and ownership? We can probably put them somewhere like SVM's integration test suite and simulate rent behavior before and after the feature is activated. What do you think?

igor56D commented on svm/src/transaction_account_state_info.rs:
```diff
@@ -1,98 +1,205 @@
                     } else {
-                        None
+                        RentState::RentExempt
                     };
```
I've confirmed it myself by checking recent snapshots + confirmed by various core devs. That said, the more correct approach here is to feature gate both the pre and post exec processing, rather than just post-exec like I'm doing now. I'll update this to avoid any gray area consensus changes, even if it's a little more convoluted.

igor56D replied:
updated: 26d3a1aea1d2a9709c1d4c1c60e67baea89f29b4

buffalojoec commented on svm/src/rent_calculator.rs:
```diff
@@ -34,22 +34,20 @@ pub enum RentState {
 /// This method has a default implementation that calls into
 /// `check_rent_state_with_account`.
 pub fn check_rent_state(
```
Depending on what you decide to do about the refactoring I suggested, this function shouldn't need to change.

buffalojoec reviewed (commented):
Thanks for adding the integration tests! I just think this diff is a bit larger than it needs to be, and this is a sensitive change. I left a few refactoring suggestions in addition to a consensus-related suggestion in the runtime.

buffalojoec reviewed (commented):
Sorry, my main refactoring comment didn't make it in my last review. I forgot to press Submit.

igor56D commented on svm/src/transaction_account_state_info.rs:
```diff
@@ -1,98 +1,221 @@
-                        ))
+pub(crate) type TransactionAccountStateInfo = Option<WritableTransactionAccountStateInfo>; // None: readonly account
+
+pub(crate) fn new_pre_exec(
```
c9e21b6c71318529e22e6f2d2cac5008312b8361

buffalojoec reviewed (commented):
This is getting super close IMO, I just still think there's a lot of churn here we can avoid. I'm down to just a few comments, then we can get this stamped from SVM side. Other SME reviewers will still need to weigh in, though.

buffalojoec reviewed (commented):
Thanks for all of the back and forth! This LGTM.

I left a comment about updating the SIMD to include the details about how to handle fee distribution in the runtime. I also want to call out that soon the runtime will also be able to credit voter rewards (inflation, block revenue) to [custom commission collector accounts](https://github.com/solana-foundation/solana-improvement-documents/blob/8f4187c7601b0ce80122a7131b45b9c01163cefb/proposals/0232-custom-commission-collector.md?plain=1#L76-L82), which must be System-owned and rent-exempt (same criteria as fee distribution here).

We should either a) update SIMD-0392 to depend on SIMD-0232 and apply the same "static account" checks to commission distribution or b) simply say that "rent exempt" as written in SIMD-0232 is replaced as a concept by SIMD-0392. What do you think?

joncinque committed:
SIMD-392: Account for rent during rewards calculation

For SIMD-392 to be complete, rewards calculation needs to ensure that
stake lamports can't be double-counted as rent and delegation.

* Add feature gate and assert during distribution
* Add lamport balances, adjust calculation
* Add tests, adjust existing tests
* Add distribution test
* Add new reward-info type
* Add new reward type everywhere, pedantically
* Add new field to CalculationEnvironment
* Update all lockfiles

joncinque commented:
https://github.com/igor56D/agave/pull/1 will include my changes in this branch, then it should be totally ready for a review

brooksprumo commented:
Wdyt about updating the PR description to point the SIMD-392 link to https://github.com/solana-foundation/solana-improvement-documents/blob/main/proposals/0392-rent-increase-adaptations.md, instead of the PR (which is now not the full SIMD)?

Edit: Resolved.

igor56D commented:
@brooksprumo good idea, will do that after I fix conflicts and CI issues.

brooksprumo commented on runtime/src/bank/fee_distribution.rs:
```diff
@@ -151,24 +152,24 @@ impl Bank {
-        if !rent_state_transition_allowed {
+            &self.rent_collector().rent,
+            self.feature_set
+                .is_active(&feature_set::relax_post_exec_min_balance_check::id()),
```
I think we can use the feature snapshot here.


```suggestion
            self.feature_set.snapshot().relax_post_exec_min_balance_check),
```

brooksprumo commented on svm/src/account_loader.rs:
```diff
@@ -383,14 +383,20 @@ pub fn validate_fee_payer(
+    let payer_post_rent_state = get_account_rent_state(
+        payer_account.lamports(),
+        payer_account.data().len(),
+        rent.minimum_balance(payer_account.data().len()),
```
Worthwhile to dedup `payer_account` data len and the minimum balance values, as they aren't changing?

brooksprumo commented on svm/src/rent_calculator.rs:
```diff
@@ -94,6 +94,44 @@ pub fn get_account_rent_state(
     }
 }
 
+/// Check whether a transition from the pre_exec_balance to the
```
Looking at the fn signature, I wasn't sure what the bool was returning. Wdyt about using "Returns" so the doc comments say "returns whether a transition [..] is valid [..]". 

```suggestion
/// Returns whether a transition from the pre_exec_balance to the
```

brooksprumo commented on svm/src/transaction_account_state_info.rs:
```diff
@@ -13,108 +15,183 @@ pub(crate) struct TransactionAccountStateInfo {
+                    let owner = *account.owner();
+
+                    let rent_state = if relax_post_exec_min_balance_check {
+                        // SIMD-0392 enabled. Assume `RentExempt` is impossible.
```
Rent exempt is impossible? Should this be rent paying?

igor56D replied:
good catch, typo

brooksprumo reviewed (commented):
Heh, there's a lot in this PR. I'm gonna need another pass for sure.

brooksprumo commented on runtime/src/bank/partitioned_epoch_rewards/distribution.rs:
```diff
@@ -739,8 +820,97 @@ mod tests {
+            .unwrap();
+        bank.store_account(&deactivating_account, &stake_account);
+        let partitioned_stake_reward = PartitionedStakeReward {
+            stake_pubkey: successful_account,
```
Should these next pubkeys be pointing at `deactivating_account` instead of `successful_account`?
```suggestion
            stake_pubkey: deactivating_account,
```

(also in the `expected_stake_reward` below too?)

joncinque replied:
It doesn't make a difference in the end, as long as both of those instances are the same, but it's certainly clearer -- done!

brooksprumo commented:
@buffalojoec - I re-requested your review. I know there are merge conflicts to resolve too. I think another look is still useful as Jon's changes have been merged into this one.

brooksprumo commented on runtime/src/inflation_rewards/mod.rs:
```diff
@@ -729,12 +793,277 @@ mod tests {
+        // -> decrease stake
+        // This case is confusing because it pays out 2 lamports in rewards,
+        // so we adjust minimum up so that even with 2 lamports in rewards, the
+        // delegation goegs down.
```
nit: typo
```suggestion
        // delegation goes down.
```

brooksprumo commented on svm/src/rent_calculator.rs:
```diff
@@ -94,6 +94,44 @@ pub fn get_account_rent_state(
+        let post_state = match (post_exec_balance, &pre_state) {
+            (0, _) => RentState::Uninitialized,
+            (post_balance, _) if post_balance >= rent_min_balance => RentState::RentExempt,
+            (post_balance, RentState::RentExempt) if post_balance >= pre_exec_balance => {
+                RentState::RentExempt
+            }
```
The codecov report indicates we don't have test coverage for this case. Seems like a useful one to add.

(there's a few others in the codecov report missing coverage as well)

igor56D replied:
there's still a few hits in the coverage report but it mostly looks spurious to me. afaict `get_inflation_reward` in rpc.rs wasn't covered in the first place anyway.

buffalojoec commented on svm/src/transaction_account_state_info.rs:
```diff
@@ -13,108 +15,183 @@ pub(crate) struct TransactionAccountStateInfo {
+    owner: Pubkey,
+}
+
+fn iter_accounts<'a>(
```
nit: I know this was my naming, but I think a better name is actually `iter_writable_accounts`.

buffalojoec commented on svm/src/transaction_account_state_info.rs:
```diff
@@ -13,108 +15,183 @@ pub(crate) struct TransactionAccountStateInfo {
+                        rent_state: get_account_rent_state(balance, data_size, min_balance),
+                        balance,
+                        data_size,
+                        owner: Pubkey::default(), // pubkey isn't needed for post-exec
```
This feels like a footgun. Even if it's not needed, I don't see a reason why we wouldn't just populate it with the real post-exec's owner. Wdyt?

buffalojoec commented on svm/src/transaction_account_state_info.rs:
```diff
@@ -13,108 +15,183 @@ pub(crate) struct TransactionAccountStateInfo {
+    fn sanitized_msg_for(key1: Pubkey, key2: Pubkey, key4: Pubkey) -> SanitizedMessage {
         let message = Message {
-            account_keys: vec![key2.pubkey(), key1.pubkey(), key4.pubkey()],
+            account_keys: vec![key2, key1, key4],
```
nit: is the idea that key1 is the program ID? Maybe that's a better name for it here?

buffalojoec commented on svm/src/transaction_account_state_info.rs:
```diff
@@ -130,36 +207,104 @@ mod test {
+            (key2, acc2),
+            (key1, AccountSharedData::default()),
+            (key3, AccountSharedData::default()),
+        ]
```
This is a little strange because the message doesn't use key3, it uses key4. I think the only reason this doesn't blow up in any tests is because keys 3 and 4 are both `AccountSharedData::default` so they have the same rent status.

igor56D replied:
now that I'm looking at this code again after a while, all these tests are very odd. I'll just do a quick refactor to clean it up.

igor56D replied:
cleaned these tests up a bit. let me know what you think.

buffalojoec commented on svm/src/transaction_account_state_info.rs:
```diff
@@ -130,36 +207,104 @@ mod test {
+    }
+
+    fn ctx_from(accounts: Vec<(Pubkey, AccountSharedData)>, rent: &Rent) -> TransactionContext<'_> {
+        TransactionContext::new(accounts, rent.clone(), 20, 20, 1)
```
nit: i don't think this parameter affects what you're testing, but I noticed you keep passing 1 as `number_of_top_level_instructions` when the messages all have two.

buffalojoec commented on svm/src/account_loader.rs:
```diff
@@ -383,14 +383,20 @@ pub fn validate_fee_payer(
+        payer_account.lamports(),
+        payer_account.data().len(),
+        rent.minimum_balance(payer_account.data().len()),
+    );
```
Aren't fee payer checks supposed to be included by SIMD-0392? It says says [right here](https://github.com/solana-foundation/solana-improvement-documents/blob/5a43c706e7bc44e8a8978b1d9186278840006940/proposals/0392-rent-increase-adaptations.md?plain=1#L118-L124):

```
transaction fee debits and subsequent fee/commission distribution are treated as
an additional balance-changing phase, and the invariant above MUST be enforced
using the pre-fee and post-fee balances.
```

igor56D replied:
yes, and I think we even discussed this earlier in this PR but it looks like something went wrong since then. Should be fixed now.

buffalojoec reviewed (commented):
I'm about halfway through this. I'll submit what I've got for the SVM side so far, and I'll circle back for the runtime side when I get another chance.

igor56D committed:
fix post-rebase compilation error

refactor: clean up account state info tests

feedback: minor touch up and add test coverage for fee distribution

brooksprumo commented:
@igor56D with the new commits/push, is the PR ready for another round of reviews?

igor56D commented:
> @igor56D with the new commits/push, is the PR ready for another round of reviews?

almost, wanted to fix conflicts and related compilation issues first but I'll be pushing up the rest of the changes. Will ping you

igor56D committed:
enforce SIMD-0392 for fee payer checks

igor56D committed:
fix post simd-232 failing test

igor56D commented:
@buffalojoec @t-nelson responded to feedback, rebased, updated, etc. Should be good for another look.

@brooksprumo still looking into adding some more test coverage but other feedback should be addressed.

The PR is getting quite large now and so I'll do another self-review as well to make sure some of the older code hasn't drifted too much.

buffalojoec commented on svm/src/transaction_account_state_info.rs:
```diff
@@ -1,120 +1,201 @@
+    owner: Pubkey,
+}
+
+fn write_iter_accounts<'a>(
```
Not to be too pedantic, but I suggested naming this `iter_writable_accounts`. The current name suggests it's performing some kind of write.

igor56D replied:
lol sorry, I need to get my eyes checked or something

igor56D replied:
7d8b88e092eaa4b9f052650f4bbdc44d0b0b4ded

joncinque commented:
Since there's an additional interaction between SIMD-0392 and SIMD-0232, I've put in #12084 just to add the feature. Let me know if you prefer some other way of handling this

joncinque commented:
https://github.com/solana-foundation/solana-improvement-documents/pull/520 for the SIMD update

buffalojoec commented on accounts-db/src/stake_rewards.rs:
```diff
@@ -139,6 +139,57 @@ impl StakeReward {
+        rent: &Rent,
+    ) -> (AccountSharedData, Self) {
+        let vote_pubkey = Pubkey::new_unique();
+        let vote_account = vote_state::create_v4_account_with_authorized(
+            &Pubkey::new_unique(),
+            &vote_pubkey,
+            [0u8; BLS_PUBLIC_KEY_COMPRESSED_SIZE],
+            &vote_pubkey,
+            1000,
+            &vote_pubkey,
+            0,
+            &vote_pubkey,
+            stake_lamports,
+        );
```
It might not matter for SIMD-0392, but we were tripping on this in https://github.com/anza-xyz/agave/pull/11877

```suggestion
        let node_pubkey = Pubkey::new_unique();
        let vote_account = vote_state::create_v4_account_with_authorized(
            &node_pubkey,
            &vote_pubkey,
            [0u8; BLS_PUBLIC_KEY_COMPRESSED_SIZE],
            &vote_pubkey,
            1000,
            &vote_pubkey,
            0,
            &node_pubkey,
            stake_lamports,
        );
```

igor56D replied:
7d8b88e092eaa4b9f052650f4bbdc44d0b0b4ded

buffalojoec commented on runtime/src/inflation_rewards/mod.rs:
```diff
@@ -111,12 +119,39 @@ fn redeem_stake_rewards<'a>(
+            if new_delegation == 0 {
+                stake.delegation.deactivation_epoch = rewarded_epoch;
+            }
+            let voter_rewards = maybe_rewards.map(|x| x.1).unwrap_or(0);
+            Some((staker_rewards, voter_rewards))
```
If `new_delegation != stake.delegation.stake`, can't these `.unwrap_or(0)` calls create a `Some(0, 0)` when there otherwise would have been a `None`? From what I can tell, this will just create a bunch of `(0, 0)` records in the stakes and commissions accumulators, so worst case we might just see them in the RPC response for `getInflationReward`.

igor56D replied:
might be worth revisiting this if/when we move to fully dynamic rent vs rare one-off adjustments.

joncinque replied:
That's correct, and the whole point of the change -- by creating a `Some(0, 0)`, we force an update to the stake account, which we'll likely need if a stake account is force-deactivated at the epoch boundary.

buffalojoec commented on runtime/src/bank/fee_distribution.rs:
```diff
@@ -191,24 +191,24 @@ impl Bank {
+            // doesn't exist yet.
+            if check_static_account_rent_state_transition(
+                pre_balance,
                 account.lamports(),
                 account.data().len(),
-            );
-            let rent_state_transition_allowed =
-                transition_allowed(&recipient_pre_rent_state, &recipient_post_rent_state);
-            if !rent_state_transition_allowed {
+                &self.rent_collector().rent,
+                0, // account index isn't relevant and only used for error message
+                feature_snapshot.relax_post_exec_min_balance_check,
+            )
+            .is_err()
+            {
```
Flagging here since this is where it looks like you'll overlap with Jon's commit from https://github.com/anza-xyz/agave/pull/12072 and https://github.com/anza-xyz/agave/pull/12084.

igor56D replied:
7d8b88e092eaa4b9f052650f4bbdc44d0b0b4ded should have all of jon's changes included now. lmk what you think.

buffalojoec commented on runtime/src/bank/partitioned_epoch_rewards/distribution.rs:
```diff
@@ -209,24 +219,43 @@ impl Bank {
-        );
+        if adjust_delegations_for_rent {
+            let minimum_balance = rent.minimum_balance(account.data().len());
+            assert!(
```
Kind of strange that these were hard panics here.

igor56D replied:
this does seem a bit sketchy. I'll replace it with a debug_assert and Err return.

igor56D replied:
7d8b88e092eaa4b9f052650f4bbdc44d0b0b4ded

buffalojoec reviewed (commented):
This is looking pretty good to me. I know there's some rebasing/merging and some test work to do, but here's my latest. Overall getting close!

igor56D committed:
feedback: improve test coverage (fee distribution), remove unnecessary asserts, and misc

igor56D commented:
Heads up: I collapse some of the commit history to make rebases and conflict resolution more manageable. Recent change history is intact, I mostly collapsed old, irrelevant commits. If anyone would like this reverted let me know, I have the history backed up.

brooksprumo commented on runtime/src/bank/partitioned_epoch_rewards/calculation.rs:
I think there's an interaction with this fn and the `stake_minimum_delegation_for_rewards` feature. If that feature is active and then rent gets raised such that a stake account no longer has the min, that stake account should be insta-deactivated. But if its min delegation is too low, it'll be filtered out here in this fn, and then will _not_ be updated (insta-deactivated) later.

brooksprumo replied:
`stake_minimum_delegation_for_rewards` will be removed by https://github.com/anza-xyz/agave/pull/12345, so that'll resolve this issue.

brooksprumo commented on runtime/src/bank/partitioned_epoch_rewards/calculation.rs:
I think there may be a corner case here too 🙃 

If a stake account that is below the new rent, but its delegated vote account is gone, we'll return None early here and then (incorrectly) miss insta-deactivating the stake account.

joncinque replied:
Ah darn, yeah that's true, nice catch!

I think we'll be ok if we return a stake reward entry here if the delegation needs to be modified. I can't think of a cleaner way to do this unfortunately, since we would need some sort of fake vote account to go any further in this function or the normal redemption logic.

If we return an entry here, we'll also create a reward entry for the vote account with 0 lamports rewarded, which seems like it won't be an issue.

cc @buffalojoec -- what do you think of this approach?

buffalojoec commented on svm/src/rent_calculator.rs:
```diff
@@ -94,6 +94,51 @@ pub fn get_account_rent_state(
+        )
+    };
+
+    if !transition_allowed(&pre_state, &post_state) {
+        let account_index = account_index as u8;
+        Err(TransactionError::InsufficientFundsForRent { account_index })
+    } else {
+        Ok(())
+    }
```
Just flagging: `check_rent_state_with_account` has a carve-out for the incinerator, while this new function doesn't.

https://github.com/anza-xyz/agave/blob/8ffefad00340fcb6a60b24c7ac942d6ff9271938/svm/src/rent_calculator.rs#L65-L66

We switched the fee payer check from using `check_rent_state_with_account` to using this function, but we assert the fee payer is not the incinerator (via signature verification) upstream during sigverify.

https://github.com/anza-xyz/agave/blob/2ae4786d20d61303f04b5b6bd1662a7065b7389f/perf/src/sigverify.rs#L46-L51

Therefore, this regression should be acceptable.

buffalojoec commented on svm/src/rent_calculator.rs:
```diff
@@ -94,6 +94,51 @@ pub fn get_account_rent_state(
+) -> TransactionResult<()> {
+    let rent_min_balance = rent.minimum_balance(data_size);
+    let (pre_state, post_state) = if relax_post_exec_min_balance_check {
+        let pre_state = if pre_exec_balance == 0 {
+            RentState::Uninitialized
+        } else {
+            RentState::RentExempt
+        };
```
I know I touched on this before in previous comments, but here we're saying that post-SIMD-0392, any non-zero balance is considered `RentExempt`. Previously, it could actually be `RentPaying`.

If we consider a fee/reward collector with balance 50 and rent minimum as 500, previously any new fees/rewards of size 50 would continue to be burned, since the account can't remain rent-paying through credits.

https://github.com/anza-xyz/agave/blob/8ffefad00340fcb6a60b24c7ac942d6ff9271938/svm/src/rent_calculator.rs#L114-L120

After this PR, the account's pre-state would register as exempt, and the fees/rewards would actually accumulate instead.

Conceptually I don't really hate this, but we are under-specified in the SIMD on this behavior.

buffalojoec replied:
The tests in `svm/src/account_loader.rs` didn't catch this regression either, which probably means we're low on coverage there as well.

igor56D replied:
The RentPaying state is effectively deprecated on MB and all relevant clusters, so tests using RentPaying accounts were intentionally modified/removed in a previous [PR](https://github.com/anza-xyz/agave/pull/9438) to make shipping 392 easier.

Once we wrap up 392 work, I'll open a SIMD to make the deprecation official.

joncinque commented on svm/src/rent_calculator.rs:
```diff
@@ -94,6 +94,51 @@ pub fn get_account_rent_state(
+        )
+    };
+
+    if !transition_allowed(&pre_state, &post_state) {
+        let account_index = account_index as u8;
+        Err(TransactionError::InsufficientFundsForRent { account_index })
+    } else {
+        Ok(())
+    }
```
nit: not a big deal, but using `transition_allowed` for both cases adds a layer of confusion, since we're kind of fudging the post state specifically with the knowledge of how `transition_allowed` works.

Up to you, but I would move this under the `else` branch, and simplify the `if relax...` branch to just return `Ok(())` in the first three match arms and `TransactionError` for the failing case

igor56D replied:
e65de78f2fe99664caf430767ed7a662505409fb I refactored this to move RentState determination logic into `get_pre_exec_account_rent_state` and `get_pre_exec_account_rent_state` but preserve the transition check. IMO this makes it clear that what's being changed isn't the transition checks but the RentState designation criteria.

buffalojoec reviewed (commented):
I spoke to @igor56D about breaking this PR up. I'll just submit what I have for now, but again I was only in SVM land for this pass.

joncinque commented on svm/tests/integration_test.rs:
```diff
@@ -2368,6 +2368,196 @@ fn simd83_account_reallocate() -> Vec<SvmTestEntry> {
+        let fee_payer_keypair = Keypair::new();
+        let fee_payer = fee_payer_keypair.pubkey();
+        let fee_payer_data =
+            mk_system_account(old_min_balance + LAMPORTS_PER_SIGNATURE, target_size);
```
If this is supposed to fall below the updated rent minimum, shouldn't the starting balance be `new_min_balance + LAMPORTS_PER_SIGNATURE - 1`?

igor56D replied:
b158afbb4bdc79b3efac4b2d0b48d28bd7fea003 fixed

joncinque commented on svm/src/transaction_account_state_info.rs:
```diff
@@ -280,27 +738,13 @@ mod test {
-                    data_size: 50,
-                }),
-            },
-            TransactionAccountStateInfo {
-                info: Some(WritableTransactionAccountStateInfo {
-                    rent_state: RentState::RentExempt,
-                    data_size: 50,
-                }),
-            },
```
nit: was this change supposed to remove the one rent exempt account? Seems like that's still good to have

igor56D replied:
good catch, we should keep the non-deleted account to verify it isn't erroneously included in the uninitialized size calculation.

igor56D replied:
b158afbb4bdc79b3efac4b2d0b48d28bd7fea003 fixed

joncinque commented on svm/src/transaction_account_state_info.rs:
```diff
@@ -193,37 +330,362 @@ mod test {
+    #[test]
+    fn test_post_exec_owner_changed_requires_full_min() {
+        let rent = Rent::default();
+        let account = Keypair::new();
+        let program = Keypair::new();
+        let other_account = Keypair::new();
+        let owner_pre = Keypair::new();
+        let owner_post = Keypair::new();
```
nit: pretty much all of these tests should use `Pubkey::new_unique()` rather than `Keypair::new()`, since there's no signing going on

igor56D replied:
b158afbb4bdc79b3efac4b2d0b48d28bd7fea003 [new PR]

joncinque reviewed (commented):
Looks good to me for the most part! Just some nits from me.

It's neat how simple the actual runtime change is in the end.
